File: php/src/VerticalAlign.php
<?php

namespace Vestaboard\Vbml;

/**
 * Port of verticalAlign.ts
 */
class VerticalAlign
{
    /**
     * @param int     $height
     * @param string  $align  'top'|'bottom'|'justified'|'center' (default)
     * @param int[][] $codes
     * @return int[][]
     */
    public static function align(int $height, string $align, array $codes): array
    {
        switch ($align) {
            case 'top':
                return $codes;

            case 'bottom':
                $reversed = array_reverse($codes);
                $result = [];
                for ($i = 0; $i < $height; $i++) {
                    $srcIndex = $height - 1 - $i;
                    $result[] = $reversed[$srcIndex] ?? [];
                }
                return $result;

            case 'justified':
                $justifiedPaddingTop = max((int)ceil(($height - count($codes)) / 2), 0);
                $result = [];
                for ($i = 0; $i < $height; $i++) {
                    $srcIndex = $i - $justifiedPaddingTop;
                    $result[] = $codes[$srcIndex] ?? [];
                }
                return $result;

            default: // center
                $paddingTop = max((int)floor(($height - count($codes)) / 2), 0);
                $result = [];
                for ($i = 0; $i < $height; $i++) {
                    $srcIndex = $i - $paddingTop;
                    $result[] = $codes[$srcIndex] ?? [];
                }
                return $result;
        }
    }
}

File: python/pyvbml/vertical_align.py
"""Vertical align.

Port of Vestaboard/vbml/src/verticalAlign.ts
"""

from __future__ import annotations

import math

from .types import Align


def vertical_align(
    height: int, align: Align, codes: list[list[int]]
) -> list[list[int]]:
    """Vertical align."""
    if align == Align.TOP:
        return codes

    if align == Align.BOTTOM:
        reversed_codes = list(reversed(codes))
        return [
            reversed_codes[height - 1 - i]
            if (height - 1 - i) < len(reversed_codes)
            else []
            for i in range(height)
        ]

    if align == Align.JUSTIFIED:
        padding_top = max(math.ceil((height - len(codes)) / 2), 0)
        return [
            codes[i - padding_top] if 0 <= i - padding_top < len(codes) else []
            for i in range(height)
        ]

    # default / center
    padding_top = max(math.floor((height - len(codes)) / 2), 0)
    return [
        codes[i - padding_top] if 0 <= i - padding_top < len(codes) else []
        for i in range(height)
    ]
